homeassistant/components/sunweg/coordinator.py

```diff
@@ -0,0 +1,100 @@
+"""The Sun WEG plant coordinator integration."""
+
+import datetime
+import logging
+
+from sunweg.api import APIHelper, LoginError, SunWegApiError
+from sunweg.plant import Plant
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryError
+from homeassistant.helpers.typing import StateType
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DEFAULT_PLANT_ID, DOMAIN, DeviceType
+
+SCAN_INTERVAL = datetime.timedelta(minutes=5)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SunWEGDataUpdateCoordinator(DataUpdateCoordinator[Plant]):
+    """SunWEG Data Update Coordinator coordinator."""
+
+    def __init__(
+        self, hass: HomeAssistant, api: APIHelper, plant_id: int, plant_name: str
+    ) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="SunWEG sensor",
+            update_interval=SCAN_INTERVAL,
+        )
+        self.api = api
+        self.plant_id = plant_id
+        self.plant_name = plant_name
+
+    async def _async_update_data(self) -> Plant:
+        """Fetch data from API endpoint."""
+        try:
+            if self.plant_id == DEFAULT_PLANT_ID:
+                # this only happens at homeassistant startup, that's why ConfigEntryError is raised in this block
```

Comment:
We have introduced a new lifecycle hook for the coordinator `_async_setup` it will be run within the `config_entry_first_refresh` function, and works with the same errors raised as `_async_update_data`